Python/optimizer.c

```diff
@@ -539,6 +549,24 @@ translate_bytecode_to_trace(
             oparg = (oparg & 0xffffff00) | executor->vm_data.oparg;
         }
 
+        if (progress_needed) {
+            progress_needed = false;
+            /* Special case the first instruction, so that we can guarantee
+            * forward progress */
+            if (opcode == JUMP_BACKWARD) {
+                instr += 2 - (int32_t)oparg;
+                continue;
+            }
+            else {
+                if (OPCODE_HAS_DEOPT(opcode)) {
+                    opcode = _PyOpcode_Deopt[opcode];
+                }
+                if (OPCODE_HAS_DEOPT(opcode)) {
+                    return 0;
```

Comment:
It might be interesting to track stats here. And definitely a level 2 DPRINTF.

Also, I see no test that exercises this line?

Reply:
Yeah, when does this actually happen?

Reply:
It can just be an assert.